### ISSUE TITLE
estargz-snapshotter support

### ISSUE THREAD

**Issue by phillebaba** [spegel-org/spegel#510]
A question was raised by @ChrisBr in #176 if Spegel would support estargz-snapshotter. As it functions differently from SOCI it will require its own analysis to determine if it is possible to use with Spegel.

**Comment by phillebaba** [comment]
Looks like the docs for mirror configuration are not up to date. However it seems like the snapshotter supports the newer CRI hosts mirror configuration. We can check the k3s PR for an example for how that would work.

https://github.com/containerd/stargz-snapshotter/issues/1647

At minimum we need to document this here as it is not documentated in stargz-snapshotter. I would like to add a test for this to verify that this is also working. The open question is whether we should add this check to verify to make sure that this configuration is set.

**Comment by phillebaba** [comment]
Still need to test this but it seems like all that needs to be changed, assuming the snapshotter is properly configured, is to change the registry config slightly.

```toml
[plugins."io.containerd.grpc.v1.cri".registry]
  config_path = "/etc/containerd/certs.d"
```

Compared to.

```toml
[plugins."io.containerd.snapshotter.v1.stargz".registry]
  config_path = "/etc/containerd/certs.d"
```

**Comment by phillebaba** [comment]
The latest stargz image uses Containerd v2 which Kind does not supporrt. We will need to wait until a full test  can be added.

https://github.com/kubernetes-sigs/kind/issues/3768

I did do some tests with an older  version of stargz and verifiedd that id does work with Spegel.

**Comment by phillebaba** [comment]
After some more testing it is a bit more complicated than expected. Some things work while others do not. The main thing being that layers are stored differently. There is however an http cache which would be interesting to look at.

https://github.com/containerd/stargz-snapshotter/blob/dbddc6d7d0921ce66a9433726b546a3fe006e27b/fs/config/config.go#L40

**Comment by simonepri** [comment]
@phillebaba any progress on this? Any idea what would be the right way to handle this?
I could find some time to contribute if we have a direction in mind already.

**Comment by phillebaba** [comment]
I have not given any attention to this feature in the last year as I have limited time.

This is a pretty big feature that I would not use myself anywhere. There are a lot of small details that aren't obvious that make the feature harder than it first seems. Mostly around how snapshotters respect mirror configuration and store layers. I have raised some of these issues with the Containerd team, but that may also take time.

**Comment by simonepri** [comment]
> I have raised some of these issues with the Containerd team, but that may also take time.

Is there some public trace of this?

How do you think the chunks should be served in the case where nodes don't have all bytes of their image (e.g. only lazy pulled part of it)? Do you think it would make sense to apply the same logic we apply to whole images to individual chunks (i.e. which node has which chunks instead of which image)? How does Spegel currently handle range requests? Any help/insight is appreciated!

For context: The use case is lazy-serving of large ML models on large GPU fleet that scale up and down and would benefit from p2p image chunks delivery.

**Comment by phillebaba** [comment]
No these were mostly in person conversations.

From my experience lazy loading is usually not the solution for speeding up ML workloads. Lazy loading works well when an image either contains files that are never accessed or files that are only needed after startup. Inference workloads tend to be a few big binaries, the driver and the model, both are needed before startup. So it would not solve your initial pull problem as the whole ML image would need to be pulled, and it would probably not make pulls faster when the image is present in the cluster.

I you want to pull chunks of an image in parallel I would suggest looking into configuring the transfer service in Containerd. This will speed up the pull speed at the cost of increased memory usage. The first pull problem is a bit trickier to solve. I built a few POCs as part of some commercial engagements so they are not OSS.

I am not against adding estargz just wanted to let you know that you may not see the value from it compared to the work it takes to add. Chunk advertisement is complicated as it would require advertising a lot more content ids. One idea I had was creating a hash with the digest and range of the content. This would make lookups easy, but at the cost of a lot more DHT queries. There are smarter ways of doing this like how IPFS does things, but it would require a bit of a rework of the pulling solution. 

I would first look and see if the estargz mirror configuration works properly and can be directed to pull from Spegel. This is usually the first challenge with custom snapshotters that you will face. Without that everything else will be for nothing.

**Comment by simonepri** [comment]
> Inference workloads tend to be a few big binaries, the driver and the model, both are needed before startup.

True, but in practice you'll always have some runtime start overhead (especially if using python), so if you can prefetch the big chunk of what you really need at t0 and keep fetching while the rest loads it can be beneficial (e.g. imagine heavy Cuda library that you need at inference time, but not at mode load time, you can start transferring weights to Cuda while the rest of the image is being pulled lazily, in parallel, under the hood).

> Lazy loading works well when an image either contains files that are never accessed or files that are only needed after startup.

Also true, and that's also the intrinsic benefit of lazy-pulling: you can stop caring about every single byte in your image, which ultimately buys you eng time that can be spent at optimizing the runtime rather than the image.

> Chunk advertisement is complicated as it would require advertising a lot more content ids.

Would it be reasonable for a node to join a ring only when it's lazy pulling has finished (at which point it can serve any range?). That would probably simplify things and would work with the current advertisement mechanism right? So IIUC the only thing to figure out then would be how to tell Spegel where the lazy-loaded image bytes are (since they are not mounted in the same way as a normal image).

**Comment by simonepri** [comment]
Just to some use-case-sepcific flavor to the conversation.

From my initial experimentation on my specific workload, these two setups have essentially the same e2e first prediction time:
 - gzip image pulling via Spegel (one warm cpu node as seed for the ring, one cold gpu node) [here single core decompression becomes a bottleneck]
 - tuned estargz image lazy-pulling via ECR (one cold gpu node)

But are both 10-20% faster than a simple (same-region) ECR pull (again for a single cold node).

Clearly the picture changes if instead of 1 node, I scale to thousands, at which point a single ECR will bottleneck. So my intuition is that combining both should bring real wins at scale (lazy-pulling eating away runtime startup overhead and Spegel providing bandwidth and reducing queries latency). I can check what's the headroom via a single in cluster mirror as a proxy of what Spegel could do.

**Comment by phillebaba** [comment]
The challenge with implementing a solution that serves "whole" images from estargz storage is that most likely the whole image will not be pulled. So we will always have a situation where partial images are present. I dont know how estargz reacts when some chunks can be mirrored while others cant. The storage part isn't really that hard. Things should not have changed since two years ago so they should be stored as fixed size chunks. Compared to SOCI estargz is simpler as SOCI does variable chunk sizes and has a different storage logic.

My guess is the reason you are seeing faster performance with estargz from ECR is most likely because of the bandwidth limits in S3. This is why transfer service in Containerd added parallel pulling. Basically many request to the same S3 bucket gives better performance than a single request. What metric are you using to measure speed? Decompression will start being the biggest limiting factor eventually.

I have heard this use case a few times. Especially when you run a bunch of ephemeral GPU nodes that need specific images that none of the long running nodes need. Using a single node as the seed works at smaller scale but gets complicated if that node is evicted or when you have a bunch of nodes requesting that image. A solution to this is [distributed prefetching and image pinning](https://drive.google.com/file/d/1wXnUkM1euAraidrCgx7f7XLwCaKt8eYx/view?usp=sharing) which uses the whole cluster to ensure all chunks are present. This is out of scope for OSS Spegel as the project focuses on serving content which is already in the cluster.

**Comment by simonepri** [comment]
> most likely the whole image will not be pulled.

AFAIU the estargz-snapshotter always pulls the entire image in background unless you enable `NoBackgroundFetch`, so eventually a node should have the whole image (and at that point in my naive solution the node could join the ring and start peering).

> A solution to this is [distributed prefetching and image pinning](https://drive.google.com/file/d/1wXnUkM1euAraidrCgx7f7XLwCaKt8eYx/view?usp=sharing) which uses the whole cluster to ensure all chunks are present

Yeah, and the interest here is essentially me exploring how to repurpose Spegel to make a lightweight/best-effort version of that. My DIY idea for this was to create a Spegel ring per image (we only have few) and a simple pool of cheap cpu-only warm nodes to seed image peering for the first GPU node, then Spegel essentially doing what you described across the cpu+gpu nodes within a ring for then next scale up events.

> What metric are you using to measure speed?

Multiple things, otel metrics within the runtime and various watchers on the k8s status to measure provisioning time, image pulling, etc for each cold-start request.

**Comment by phillebaba** [comment]
I am fine with adding an initial support for estargz snapshotter that only works when the full image is pulled. That way the problem of partial chunks can be solved at a later date. We would need to expand the Containerd integration tests, but I can assist with that when it gets there.

**Comment by simonepri** [comment]
@phillebaba ok I think I got the setup working e2e.

I sent you the 2 changes I think we need to make eStargz being able to pull from Spegel peers.
Once those are in I have an integration test that proves eStargz -> Spegel peer pulling works. Then the last bit is then enabling Spegel to serve eStargz cached image blobs (of which I have a first version but happy to iterate on it with you)

**Comment by simonepri** [comment]
@phillebaba ok looks like we cleared the two of the three blockers for estargz support. 🥳 

My plan would now be to send you a PR for a estargz integration tests to cover what is working/not working and then a final PR to fix the serving part (assuming you are not fixing that already with your refactorings).
